File: packages/arcis-node/src/sanitizers/encode.ts
/**
 * @module @arcis/node/sanitizers/encode
 * Context-aware output encoding for XSS prevention.
 *
 * Wrong-context encoding is the #1 cause of XSS bypasses in "protected" apps.
 * A single sanitize() is not enough when output goes to JS, CSS, or attribute contexts.
 */

// HTML entity map — covers the 5 dangerous chars in HTML body context
const HTML_ENTITIES: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#x27;',
};

const HTML_ENCODE_RE = /[&<>"']/g;

/**
 * Encodes for HTML body context. Entity-encodes & < > " '
 *
 * Use when outputting to HTML element content:
 *   `<p>${encodeForHtml(userInput)}</p>`
 */
export function encodeForHtml(value: string): string {
  if (!value) return '';
  return value.replace(HTML_ENCODE_RE, (ch) => HTML_ENTITIES[ch]);
}

/**
 * Encodes for HTML attribute context.
 * All non-alphanumeric characters are encoded as `&#xHH;` hex entities.
 *
 * Use when outputting to HTML attributes:
 *   `<div title="${encodeForAttribute(userInput)}">`
 */
export function encodeForAttribute(value: string): string {
  if (!value) return '';
  let result = '';
  for (let i = 0; i < value.length; i++) {
    const ch = value.charCodeAt(i);
    // Allow a-z A-Z 0-9
    if (
      (ch >= 0x30 && ch <= 0x39) || // 0-9
      (ch >= 0x41 && ch <= 0x5a) || // A-Z
      (ch >= 0x61 && ch <= 0x7a)    // a-z
    ) {
      result += value[i];
    } else {
      result += `&#x${ch.toString(16).toUpperCase()};`;
    }
  }
  return result;
}

/**
 * Encodes for JavaScript string context.
 * Non-alphanumeric characters are escaped as `\xHH` (ASCII) or `\uHHHH` (Unicode).
 *
 * Use when embedding in JS string literals:
 *   `var x = '${encodeForJs(userInput)}';`
 */
export function encodeForJs(value: string): string {
  if (!value) return '';
  let result = '';
  for (let i = 0; i < value.length; i++) {
    const ch = value.charCodeAt(i);
    // Allow a-z A-Z 0-9
    if (
      (ch >= 0x30 && ch <= 0x39) || // 0-9
      (ch >= 0x41 && ch <= 0x5a) || // A-Z
      (ch >= 0x61 && ch <= 0x7a)    // a-z
    ) {
      result += value[i];
    } else if (ch < 0x100) {
      result += `\\x${ch.toString(16).toUpperCase().padStart(2, '0')}`;
    } else {
      result += `\\u${ch.toString(16).toUpperCase().padStart(4, '0')}`;
    }
  }
  return result;
}

/**
 * Encodes for URL parameter context. Percent-encodes all non-unreserved chars.
 *
 * Use when building query strings:
 *   `?q=${encodeForUrl(userInput)}`
 */
export function encodeForUrl(value: string): string {
  if (!value) return '';
  // encodeURIComponent handles most cases but doesn't encode: ! ' ( ) *
  // We encode these additionally for full safety per RFC 3986
  return encodeURIComponent(value).replace(/[!'()*]/g, (ch) => {
    return `%${ch.charCodeAt(0).toString(16).toUpperCase()}`;
  });
}

/**
 * Encodes for CSS value context.
 * Non-alphanumeric characters are hex-escaped as `\HH ` (trailing space per CSS spec).
 *
 * Use when embedding in CSS values:
 *   `content: '${encodeForCss(userInput)}';`
 */
export function encodeForCss(value: string): string {
  if (!value) return '';
  let result = '';
  for (let i = 0; i < value.length; i++) {
    const ch = value.charCodeAt(i);
    // Allow a-z A-Z 0-9
    if (
      (ch >= 0x30 && ch <= 0x39) || // 0-9
      (ch >= 0x41 && ch <= 0x5a) || // A-Z
      (ch >= 0x61 && ch <= 0x7a)    // a-z
    ) {
      result += value[i];
    } else {
      // CSS hex escape: backslash + hex code + trailing space (CSS spec requirement)
      result += `\\${ch.toString(16).toUpperCase()} `;
    }
  }
  return result;
}

File: packages/arcis-python/arcis/sanitizers/encode.py
"""
Context-aware output encoding for XSS prevention.

Wrong-context encoding is the #1 cause of XSS bypasses in "protected" apps.
A single sanitize() is not enough when output goes to JS, CSS, or attribute contexts.
"""

from urllib.parse import quote as _url_quote


# HTML entity map — covers the 5 dangerous chars in HTML body context
_HTML_ENTITIES = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#x27;",
}


def _is_alphanumeric(ch: str) -> bool:
    """Check if a character is ASCII alphanumeric."""
    return ch.isascii() and ch.isalnum()


def encode_for_html(value: str) -> str:
    """Encode for HTML body context. Entity-encodes & < > " '

    Use when outputting to HTML element content::

        f"<p>{encode_for_html(user_input)}</p>"
    """
    if not value:
        return ""
    return "".join(_HTML_ENTITIES.get(ch, ch) for ch in value)


def encode_for_attribute(value: str) -> str:
    """Encode for HTML attribute context.

    All non-alphanumeric characters are encoded as ``&#xHH;`` hex entities.

    Use when outputting to HTML attributes::

        f'<div title="{encode_for_attribute(user_input)}">'
    """
    if not value:
        return ""
    result = []
    for ch in value:
        if _is_alphanumeric(ch):
            result.append(ch)
        else:
            result.append(f"&#x{ord(ch):X};")
    return "".join(result)


def encode_for_js(value: str) -> str:
    r"""Encode for JavaScript string context.

    Non-alphanumeric characters are escaped as ``\xHH`` (ASCII) or ``\uHHHH`` (Unicode).

    Use when embedding in JS string literals::

        f"var x = '{encode_for_js(user_input)}';"
    """
    if not value:
        return ""
    result = []
    for ch in value:
        code = ord(ch)
        if _is_alphanumeric(ch):
            result.append(ch)
        elif code < 0x100:
            result.append(f"\\x{code:02X}")
        else:
            result.append(f"\\u{code:04X}")
    return "".join(result)


def encode_for_url(value: str) -> str:
    """Encode for URL parameter context. Percent-encodes all non-unreserved chars.

    Use when building query strings::

        f"?q={encode_for_url(user_input)}"
    """
    if not value:
        return ""
    # quote with empty safe= encodes everything except unreserved chars (RFC 3986)
    return _url_quote(value, safe="")


def encode_for_css(value: str) -> str:
    r"""Encode for CSS value context.

    Non-alphanumeric characters are hex-escaped as ``\HH `` (trailing space per CSS spec).

    Use when embedding in CSS values::

        f"content: '{encode_for_css(user_input)}';"
    """
    if not value:
        return ""
    result = []
    for ch in value:
        if _is_alphanumeric(ch):
            result.append(ch)
        else:
            # CSS hex escape: backslash + hex code + trailing space
            result.append(f"\\{ord(ch):X} ")
    return "".join(result)
